ethers/contract.nim

```diff
@@ -97,6 +101,13 @@ func isConstant(procedure: NimNode): bool =
       return true
   false
 
+func isTxResponse(returntype: NimNode): bool =
+  return returntype.kind == nnkPrefix and
+    returntype[0].kind == nnkIdent and
+    returntype[0].strVal == "?" and
```

Comment:
Can we define a custom return type and check for that instead?

Reply:
> Can we define a custom return type and check for that instead?

Great idea. Have added `Waitable` as the type alias for `?TransactionResponse`.

The only down side I can see from this is that if a user returns `Waitable` but does not chain the `.wait` afterwards, they end up with `Option[TransactionResponse]`, but it is not obvious that `Waitable` is an option. However, it is a small concession, and the compiler will pick it up.

Reply:
I actually liked the return type `?TransactionResponse` :sweat_smile: 

What I meant by my remark is that I would prefer it if the code did a semantic check on the type, instead of a syntactic check. So instead of checking whether the type is written as `?`, you could check whether the type is equivalent to `Option[TransactionResponse]`. So that it wouldn't matter whether the user typed `Option[TransactionResponse]`, `?TransactionResponse`, or even `MyCustomType`, where `MyCustomType` is defined as `Option[TransactionResponse]`.

Reply:
Definitely agreed. I am likely wrong, but I did try `returntype.getType` and `returntype.getTypeImpl`, and the compiler complained that the node has no type. Perhaps the compiler hasn't done the lookup yet and all we have is the AST (which shows a `nnkPrefix` and two `Ident` children: `?` and `TransactionResponse`)?

Reply:
Right, I was afraid that that might be the case. We can go with `Waitable` for now then, and see whether we can solve this at a later stage.

Reply:
Getting types from untyped macros procedure is fraught with peril (gets even worst with generics etc)
If I may suggest, it's sometime best to instead inject a at the beggining of the body
```nim
static: checkReturnType(type(result))
```
with the macro. Not clean, but works

Reply:
Nice, that would probably solve it @Menduist!

Reply:
@Menduist, i've added
```nim
static: checkReturnType(type(result))
```
to body and removed the check from inside the contract macro, but `checkReturnType` is never called. The AST of `mint`, for example, looks like this (with the static stmt):
```
ProcDef
  Ident "mint"
  Empty
  Empty
  FormalParams
    Empty
    IdentDefs
      Ident "token"
      Ident "TestToken"
      Empty
    IdentDefs
      Ident "holder"
      Ident "Address"
      Empty
    IdentDefs
      Ident "amount"
      Ident "UInt256"
      Empty
  Pragma
    Ident "view"
  StaticStmt
    StmtList
      Call
        Ident "checkReturnType"
        Call
          Sym "type"
          Ident "result"
  Command
    Sym "await"
    Call
      OpenSymChoice
        Sym "call"
        Sym "call"
        Sym "call"
      Ident "token"
      StrLit "mint"
      TupleConstr
        Ident "holder"
        Ident "amount"
```

Reply:
I added other func signatures too in the hopes of it being called, eg:
```nim
func checkReturnType*(T: type) =
func checkReturnType*[T]() =
```
But none seem to work.